File: Wdtc/src/main/java/org/wdt/wdtc/ui/window/ModChooseVersionWindow.java
package org.wdt.wdtc.ui.window;

import com.jfoenix.controls.JFXButton;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import org.wdt.wdtc.core.download.fabric.FabricAPIDownloadTask;
import org.wdt.wdtc.core.download.fabric.FabricAPIVersionList;
import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo;
import org.wdt.wdtc.core.download.fabric.FabricVersionList;
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo;
import org.wdt.wdtc.core.download.forge.ForgeVersionList;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.download.infterface.VersionListInterface;
import org.wdt.wdtc.core.download.quilt.QuiltInstallTask;
import org.wdt.wdtc.core.download.quilt.QuiltVersionList;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.utils.ModUtils;

import java.io.IOException;

public class ModChooseVersionWindow {
  private final ModUtils.KindOfMod kind;

  private final Stage MainStage;
  private final WindwosSizeManger size;
  private final Launcher launcher;

  public ModChooseVersionWindow(ModUtils.KindOfMod kind, Stage MainStage, Launcher launcher) {
    this.kind = kind;
    this.MainStage = MainStage;
    this.launcher = launcher;
    this.size = new WindwosSizeManger(MainStage);
  }

  public VersionListInterface getModVersionList() throws IOException {
    if (kind == ModUtils.KindOfMod.FORGE) {
      return new ForgeVersionList(launcher);
    } else if (kind == ModUtils.KindOfMod.FABRICAPI) {
      return new FabricAPIVersionList(launcher);
    } else if (kind == ModUtils.KindOfMod.QUILT) {
      return new QuiltVersionList(launcher);
    } else {
      return new FabricVersionList();
    }
  }

  public void setModChooser() throws IOException {
    JFXButton back = new JFXButton("返回");
    back.getStyleClass().add("BlackBorder");
    Pane pane = new Pane();
    VBox ButtonList = new VBox();
    ScrollPane list = new ScrollPane();
    Label tips = new Label("选择一个Mod版本:");
    tips.setLayoutX(149.0);
    tips.setLayoutY(67.0);
    list.setLayoutY(134.0);
    list.setPrefSize(600, 316);
    Platform.runLater(() -> {
      try {
        for (VersionJsonObjectInterface versionJsonObject : getModVersionList().getVersionList()) {
          JFXButton VersionButton = getVersionButton(versionJsonObject, ButtonList);
          size.ModifyWindwosSize(ButtonList, VersionButton);
        }
      } catch (IOException e) {
        ExceptionWindow.setErrorWin(e);
      }
    });
    list.setContent(ButtonList);
    size.ModifyWindwosSize(pane, list, back, tips);
    pane.setBackground(Consoler.getBackground());
    Consoler.setStylesheets(pane);
    MainStage.setScene(new Scene(pane));
    back.setOnAction(event -> {
      ModChooseWindow Choose = new ModChooseWindow(launcher, MainStage);
      Choose.setChooseWin();
    });

  }

  private JFXButton getVersionButton(VersionJsonObjectInterface versionJsonObject, VBox ButtonList) {
    JFXButton VersionButton = new JFXButton(versionJsonObject.getVersionNumber());
    VersionButton.setStyle("-fx-border-color: #000000");
    VersionButton.setPrefWidth(600.0);
    VersionButton.setOnAction(event -> {
      switch (kind) {
        case FORGE -> launcher.setForgeModDownloadInfo(new ForgeDownloadInfo(launcher, versionJsonObject));
        case FABRIC -> launcher.setFabricModInstallInfo(new FabricDonwloadInfo(launcher, versionJsonObject));
        case QUILT -> launcher.setQuiltModDownloadInfo(new QuiltInstallTask(launcher, versionJsonObject));
        case FABRICAPI -> {
          if (launcher.fabricModInstallInfo != null) {
            launcher.fabricModInstallInfo.setApiDownloadTask(new FabricAPIDownloadTask(launcher, versionJsonObject));
          }
        }
      }
      ModChooseWindow Choose = new ModChooseWindow(launcher, MainStage);
      Choose.setChooseWin();
      ButtonList.getChildren().clear();
    });
    return VersionButton;
  }
}

File: Wdtc/src/main/kotlin/org/wdt/wdtc/ui/window/ModChooseVersionWindow.kt
package org.wdt.wdtc.ui.window

import com.jfoenix.controls.JFXButton
import javafx.event.EventHandler
import javafx.scene.Scene
import javafx.scene.control.Label
import javafx.scene.control.ScrollPane
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.VBox
import javafx.stage.Stage
import kotlinx.coroutines.*
import org.wdt.wdtc.core.download.fabric.FabricAPIDownloadTask
import org.wdt.wdtc.core.download.fabric.FabricAPIVersionList
import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo
import org.wdt.wdtc.core.download.fabric.FabricVersionList
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo
import org.wdt.wdtc.core.download.forge.ForgeVersionList
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.download.quilt.QuiltInstallTask
import org.wdt.wdtc.core.download.quilt.QuiltVersionList
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.ckeckIsNull
import java.io.IOException

class ModChooseVersionWindow(
  private val kind: KindOfMod,
  private val mainStage: Stage,
  private val version: Version
) {
  private val size = mainStage.getSizeManger()

  private val modVersionList
    get() = when (kind) {
      KindOfMod.FORGE -> {
        ForgeVersionList(version).versionList
      }

      KindOfMod.FABRICAPI -> {
        FabricAPIVersionList(version).versionList
      }

      KindOfMod.QUILT -> {
        QuiltVersionList(version).versionList
      }

      else -> {
        FabricVersionList().versionList
      }
    }


  @Throws(IOException::class)
  fun setModChooser() {
    val back = JFXButton("返回").apply {
      styleClass.add("BlackBorder")
      onAction = EventHandler {
        ModChooseWindow(version, mainStage).run {
          setChooseWin()
        }
      }
    }
    val buttonList = VBox()
    val tips = Label("选择一个Mod版本:").apply {
      layoutX = 130.0
      layoutY = 20.0
    }
    try {
      CoroutineScope(Dispatchers.Unconfined).launch {
        modVersionList.forEach {
          size.modifyWindwosSize(buttonList, getVersionButton(it, buttonList))
        }
      }
    } catch (e: IOException) {
      setErrorWin(e)
    }
    val list = ScrollPane().apply {
      setTopAnchor(40.0)
      setRightAnchor(0.0)
      setLeftAnchor(0.0)
      setBottomAnchor(0.0)
      content = buttonList
    }
    AnchorPane().apply {
      background = wdtcBackground
      setStylesheets()
    }.let {
      size.modifyWindwosSize(it, list, back, tips)
      mainStage.setScene(Scene(it))
    }
  }

  private fun getVersionButton(versionJsonObject: VersionsJsonObjectInterface, buttonList: VBox): JFXButton {
    return JFXButton(versionJsonObject.versionNumber).apply {
      style = "-fx-border-color: #000000"
      prefWidth = 580.0
      onAction = EventHandler {
        when (kind) {
          KindOfMod.FORGE -> version.forgeModDownloadInfo = ForgeDownloadInfo(version, versionJsonObject)
          KindOfMod.FABRIC -> version.fabricModInstallInfo = FabricDonwloadInfo(version, versionJsonObject)
          KindOfMod.QUILT -> version.quiltModDownloadInfo = QuiltInstallTask(version, versionJsonObject)
          KindOfMod.FABRICAPI -> version.fabricModInstallInfo.ckeckIsNull().apiDownloadTask =
            FabricAPIDownloadTask(version, versionJsonObject)

          else -> {}
        }
        ModChooseWindow(version, mainStage).run {
          setChooseWin()
        }
        buttonList.children.clear()
      }
    }
  }
}
